From 032c0ddead2f7acb5be56a0577159e03b3d4dc69 Mon Sep 17 00:00:00 2001
From: Tobias Stoeckmann <tobias@stoeckmann.org>
Date: Fri, 16 Oct 2026 17:03:40 +0200
Subject: [PATCH] uu: Do not handle non-ASCII lines anymore

The filter becomes too eager and can start spotting uu encoded files
within archives.

Effectively reverts #3456

Signed-off-by: Tobias Stoeckmann <tobias@stoeckmann.org>

(cherry picked from commit 032c0ddead2f7acb5be56a0577159e03b3d4dc69)
---
 libarchive/archive_read_support_filter_uu.c | 11 +++++++++++
 libarchive/test/test_read_filter_uudecode.c |  1 -
 2 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/libarchive/archive_read_support_filter_uu.c b/libarchive/archive_read_support_filter_uu.c
index 4ec2a29..26b0929 100644
--- a/libarchive/archive_read_support_filter_uu.c
+++ b/libarchive/archive_read_support_filter_uu.c
@@ -232,6 +232,17 @@ read_line_ahead(struct archive_read_filter *f, ssize_t parsed,
 			line = b + start;
 			*len = p - line;
 			*unconsumed = *len + nl;
+
+			/* Only ASCII lines are valid. */
+			while (p > line) {
+				unsigned char c = *--p;
+
+				if (c < 0x20 || c > 0x7F) {
+					*len = 0;
+					*unconsumed = 0;
+					return (NULL);
+				}
+			}
 		} else {
 			/* Keep offset for check of possibly missing \n in \r\n. */
 			if (step != 1)
diff --git a/libarchive/test/test_read_filter_uudecode.c b/libarchive/test/test_read_filter_uudecode.c
index f6bbaf6..88930a0 100644
--- a/libarchive/test/test_read_filter_uudecode.c
+++ b/libarchive/test/test_read_filter_uudecode.c
@@ -41,7 +41,6 @@ static const char empty_archive[] = {
 };
 
 static const char archive64[] = {
-"UTF-8 grinning face with smiling eyes \xF0\x9F\x98\x81\n"
 "begin-base64 644 test_read_uu.Z\n"
 "H52QLgAIHEiwoMGDCBMqXMiwIUIYEG/UqAECAMQYEmFUvJhxI8SPIDXGgDFjBg0YNDDOsAECxsga\n"
 "NmIAAFHDoc2bOHPqBFBnDp0wcizCoJOmzc6ERI0ePRhSo1CQFZdKnUq1qtWrWLNq3cq1q9evYMOK\n"
-- 
2.53.0

